cirq-core/cirq/devices/named_topologies.py

```diff
@@ -0,0 +1,337 @@
+
+        extra_w = self.width % 2
+        if extra_w:
+            # Add extra width to the final half-column
+            for unit_row in range(unit_cell_height):
+                _add_edge(unit_row, unit_cell_width, which=0)
+                _add_edge(unit_row, unit_cell_width, which=2)
+
+        if extra_w and extra_h:
+            # Add the final corner node when we have both ragged boundaries
+            _add_edge(unit_cell_height, unit_cell_width, which=0)
+
+        object.__setattr__(self, 'graph', g)
+
+        # The number of edges = width * height (see unit tests). This can be seen if you remove
+        # all vertices and replace edges with dots.
+        # The formula for the number of vertices is not that nice, but you can derive it by
+        # summing big and small Xes in the asciiart in the docstring.
+        # There are (width//2 + 1) * (height//2 + 1) small xes and
+        # ((width + 1)//2) * ((height + 1)//2) big ones.
+        n_nodes = (self.width // 2 + 1) * (self.height // 2 + 1)
+        n_nodes += ((self.width + 1) // 2) * ((self.height + 1) // 2)
+        object.__setattr__(self, 'n_nodes', n_nodes)
+
+    def draw(self, ax=None, tilted=True, **kwargs):
+        """Draw this graph
+
+        Args:
+            ax: Optional matplotlib axis to use for drawing.
+            tilted: If True, directly position as (row, column); otherwise,
+                rotate 45 degrees to accommodate the diagonal nature of this topology.
+            kwargs: Additional arguments to pass to `nx.draw_networkx`.
+        """
+        return draw_gridlike(self.graph, ax=ax, tilted=tilted, **kwargs)
+
+    def nodes_as_gridqubits(self) -> List['cirq.GridQubit']:
+        """Get the graph nodes as cirq.GridQubit"""
+        return [GridQubit(r, c) for r, c in sorted(self.graph.nodes)]
+
+    def _json_dict_(self) -> Dict[str, Any]:
+        return dataclass_json_dict(self)
+
+
+def get_placements(
+    big_graph: nx.Graph, small_graph: nx.Graph, max_placements=100_000
+) -> List[Dict]:
+    """Get 'placements' mapping small_graph nodes onto those of `big_graph`.
+
+    We often consider the case where `big_graph` is a nx.Graph representation of a Device
+    whose nodes are `cirq.Qid`s like `GridQubit`s and `small_graph` is a NamedTopology graph.
+    In this case, this function returns a list of placement dictionaries. Each dictionary
+    maps the nodes in `small_graph` to nodes in `big_graph` with a monomorphic relationship.
+    That's to say: if an edge exists in `small_graph` between two nodes, it will exist in
+    `big_graph` between the mapped nodes.
+
+    We restrict only to unique set of `big_graph` qubits. Some monomorphisms may be basically
+    the same mapping just rotated/flipped which we purposefully exclude. This could
+    exclude meaningful differences like using the same qubits but having the edges assigned
+    differently, but it prevents the number of placements from blowing up.
+    """
```

Comment:
Refactored into Args, Returns. Tightened up some of the wording. Kept most of the content. I think it's all pretty important. 

This is in a class of functions that doesn't correspond to the super pure general everything function: instead, it encodes general ideas I usually need when doing the particular form of graph matching for putting a circuit on a device. It might not be for everyone. The pure, general version would just be networkx's `sugraph_monomorphism_iter()`. The intricacies in this version need to be documented *and motivated*